test/syscalls/linux/mmap.cc

```diff
@@ -580,16 +580,16 @@ const uint8_t machine_code[] = {
 
 // PROT_EXEC allows code execution
 TEST_F(MMapTest, ProtExec) {
-  uintptr_t addr;
+  std::atomic<uintptr_t> addr;
   uint32_t (*func)(void);
 
   EXPECT_THAT(addr = Map(0, kPageSize, PROT_EXEC | PROT_READ | PROT_WRITE,
                          MAP_PRIVATE | MAP_ANONYMOUS, -1, 0),
               SyscallSucceeds());
 
-  memcpy(reinterpret_cast<void*>(addr), machine_code, sizeof(machine_code));
+  memcpy(reinterpret_cast<void*>(addr.load(std::memory_order_acquire)), machine_code, sizeof(machine_code));
 
```

Comment:
The memory barrier here seems to be incidentally solving the issue.

Maybe change the permissions explicitly, forcing whatever flush might be needed? Then you're guaranteed the problem will be solved. (The system call itself is a memory barrier, but the remap is obviously much more.)

To me this seems more much "obviously correct" and self-documenting. E.g.

EXPECT_THAT(addr = Map(..... PROT_READ | PROT_WRITE) ...);

memcpy(...);

EXPECT_THAT(Map(addr, ...., PROT_EXEC | PROT_READ);

func = reinterpret_cast<....;

Reply:
(FWIW, I think only the second barrier is probably necessary here, as the stores from the memcpy have probably been reordered around the func() call. But the argument about "obviously correct" I think still holds and we should just do two explicit Map calls with appropriate permissions each time.)

Reply:
Hi,
I have changed the addr in memcpy() to memory_order_relaxed.
As you see in my PR description, we only need to add the memory barrier after the memcpy.
'''
 The code to directly use the memory barrier is as follows:
  memcpy(reinterpret_cast<void*>(addr), machine_code, sizeof(machine_code));
  ############MEMORY BARRIER#######
  isb()
  #################################
  func = reinterpret_cast<uint32_t (*)(void)>(addr);

So, I only used the memory_order_acquire after the memcpy in the latest modification.
Do you think such a modification is ok?

Thanks.

Reply:
I'm good with either. I just think the two Maps is more obvious to the reader. If you want to keep this one, please add a quick comment that indicates the barrier is required on ARM, so that someone doesn't remove it accidentally :) Then I will merge.

Reply:
Hi @amscanne 
Done.